### PR TITLE
flow fixes

### PR DESCRIPTION
**Pull request by jquense**
Still not correct, since the host config expects all things implemented. This is a great example of what I was talking about, instead of implementing controlled components i spent two hours configuring flow and ran out of time

### CODE REVIEW

**Comment by prometheansacrifice** [comment]
What was the flow error? Is this branched off the flow branch? Last I checked there were no flow errors there.

**Comment by jquense** [comment]
Your PR didn't have a flow pragma on the reconciler types, so it wasn't being typed checked at all

**Comment by prometheansacrifice** [comment]
@jquense So flow wasn't type checking HostConfig at?
I skipped the pragma on purpose. That file is a stopgap fix. Those types are supposed to come from upstream. Flow checking that file would be like unit testing library code. 

I did try passing wrong types to the HostConfig object. Flow did complain. Was it not complaining for you?

**Comment by jquense** [comment]
So if the file isn't checked it's the same as if the types weren't there. The HostConfig type was `any`, which defeats most of the purpose of having it, since weren't getting any type safety in index.js

Generally the stopgap solution for missing types in dependencies is a lib def file in flow-typed/ I started to put one together, but it's not complete

**Comment by prometheansacrifice** [comment]
[HostConfig was not annotated any](https://github.com/jquense/react-dom-lite/blob/master/src/index.js#L15)

**Comment by jquense** [comment]
Sorry, I know that. But flow could not determine the type because the type file was broken, so regardless of what the annotation said. Flow reported the value as `any`. Do you have editor support for flow type? You usually can inspect the type inference flow or turn on flow coverage you'd see that the types were not doing what we'd want

**Comment by jquense** [comment]
Sorry if I'm not not doing a good job explaining the issue!

**Comment by prometheansacrifice** [comment]
Hey, not sure if this helps. Nevertheless,  I put together these types in [gist](https://gist.github.com/prometheansacrifice/6e470f6dd52710af915de1b240f5bacd) You seem to have done more work though.

**Comment by jquense** [comment]
should be good to go now

**Comment by jquense** [comment]
this look good for now?

**Comment by prometheansacrifice** [comment]
I added some review comments.

File: src/index.js [hunk 1]
```diff
@@ -124,37 +158,38 @@ const HostConfig: HC = {
...
 
 DOMLiteRenderer.injectIntoDevTools({
-  bundleType: 1, // 0 for PROD, 1 for DEV
-  version: '0.1.0', // version for your renderer
-  rendererPackageName: 'custom-renderer', // package name
-  findHostInstanceByFiber: DOMLiteRenderer.findHostInstance // host instance (root)
+  bundleType: __DEV__ ? 1 : 0,
+  version: '0.1.0',
+  rendererPackageName: 'react-dom-lite',
+  findFiberByHostInstance: DOMLiteRenderer.findHostInstance
```
Comment (gaearon) [inline src/index.js]: This doesn't look right.

File: src/index.js [hunk 2]
```diff
@@ -89,33 +103,53 @@ const HostConfig: HC = {
...
-  now: () => {},
+  now() {
+    return typeof performance === 'object' &&
+      typeof performance.now === 'function'
+      ? performance.now()
+      : Date.now();
+  },
 
   useSyncScheduling: true,
+  scheduleDeferredCallback: window.requestIdleCallback,
```
[the remark this answers is not in the document]
Comment (jquense) [reply src/index.js]: good catch, I wasn't geearing up to handle SSR at the moment so this seemed ok as a way to keep flow happy until we extract it out properly

File: flow-typed/react-reconciler.js [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    value: V
+  }
+};
+
+type ReactPortal = {
+  $$typeof: Symbol | number,
+  key: null | string,
+  containerInfo: any,
+  children: ReactNodeList,
+  // TODO: figure out the API for cross-renderer implementation.
```
[the remark this answers is not in the document]
Comment (jquense) [reply flow-typed/react-reconciler.js]: Not intentionally persay, but i was trying to touch the files as little as possible so it's easier to move upstream (granted comments wouldn't affect that!)

File: flow-typed/react-reconciler.js [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+  implementation: any
+};
+
+declare module 'react-reconciler' {
+  declare export type OpaqueHandle = Fiber;
+  declare export type OpaqueRoot = FiberRoot;
+  declare export type ExpirationTime = number;
+
+  declare export type HostConfig<T, P, I, TI, HI, PI, C, CC, CX, PL> = {
```
Comment (jquense) [inline flow-typed/react-reconciler.js]: "how many layers of generics you on?",
"like 4 or five my dude"
"You're like a little baby. watch this"

@gaearon single letter generic types seem fairly commmon in Flow. Is there a reason, that you know of, why slightly more descriptive convention isn't used, like `<TInstance>` or something a la C# ?

File: src/index.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (prometheansacrifice) [reply src/index.js]: Not a blocker: we can patch it later

File: flow-typed/react-reconciler.js (hunk 4 above)
Comment (gaearon) [reply flow-typed/react-reconciler.js]: I dunno, that's what I'd prefer too but @sebmarkbage didn't like it (I don't remember why).

File: src/index.js [hunk 5]
```diff
@@ -1,50 +1,81 @@
...
+  string, // T: component type
+  Props, // P: props
+  Element, // I: component instance
+  Text, // TI: component text instance
+  Element, // HI: Hydration Instance
+  Element | Text, // PI: Public instance
+  Element, // C: Container instance
+  any, // Child container instance
+  HostContext, // CX: Host context
+  Array<[string, any]> // PL: prepare update result
```
Comment (gaearon) [inline src/index.js]: FWIW Flow should be able to infer this. Although personally I do prefer an explicit annotation myself.

**Pull request merged**